File: src/GCov/Data/CoverageThread.java
package GCov.Data;

import GCov.Notification.GCovNotification;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

public class CoverageThread extends Thread {

    private static final Logger log = Logger.getInstance(CoverageData.class);
    private final Runnable m_runner;
    private final String m_buildDirectory;
    private final Project m_project;
    private final CoverageData m_data;

    public CoverageThread(@NotNull Project project, String buildDirectory, Runnable runner) {
        m_runner = runner;
        m_buildDirectory = buildDirectory;
        m_project = project;
        m_data = CoverageData.getInstance(m_project);
    }

    private boolean generateGCDA(@NotNull Path gcda) {
        int retCode = -1;
        try {
            String nullFile;
            if (System.getProperty("os.name").startsWith("Windows")) {
                nullFile = "NUL";
            } else {
                nullFile = "/dev/zero";
            }
            ProcessBuilder builder = new ProcessBuilder("gcov", "-i", "-m", "-b", gcda.toString())
                    .directory(gcda.toFile().getParentFile())
                    .redirectOutput(new File(nullFile)).redirectErrorStream(true);
            Process p = builder.start();
            try {
                retCode = p.waitFor();
            } catch (InterruptedException e) {
                Notification notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                        .createNotification("Process timed out", NotificationType.ERROR);
                Notifications.Bus.notify(notification,m_project);
            }
            if (retCode != 0) {
                Notification notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                        .createNotification("\"gcov\" returned with error code " + retCode, NotificationType.ERROR);
                Notifications.Bus.notify(notification,m_project);
            }
        } catch (IOException e) {
            Notification notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                    .createNotification("\"gcov\" was not found in system path", NotificationType.ERROR);
            Notifications.Bus.notify(notification,m_project);
        }
        return retCode == 0;
    }

    private void parseGCov(@NotNull List<String> lines) {
        CoverageFileData currentFile = null;
        for (String line : lines) {
            String substring = line.substring(line.indexOf(':') + 1);
            switch (line.substring(0, line.indexOf(':'))) {
                case "file": {
                    currentFile = m_data.getData().get(substring);
                    if (currentFile == null) {
                        currentFile = new CoverageFileData(substring,
                                m_project.getBasePath() != null && substring.startsWith(m_project.getBasePath()));
                        m_data.getData().put(substring, currentFile);
                    }
                }
                break;
                case "function": {
                    log.assertTrue(currentFile != null, "\"function\" statement found before a file statement");
                    String data = substring;
                    int startLine = Integer.parseInt(data.substring(0, data.indexOf(',')));
                    data = data.substring(data.indexOf(',') + 1);
                    int endLine = Integer.parseInt(data.substring(0, data.indexOf(',')));
                    data = data.substring(data.indexOf(',') + 1);
                    int executionCount = Integer.parseInt(data.substring(0, data.indexOf(',')));
                    data = data.substring(data.indexOf(',') + 1);
                    currentFile.emplaceFunction(startLine, endLine, executionCount, data);
                }
                break;
                case "lcount": {
                    log.assertTrue(currentFile != null, "\"lcount\" statement found before a file statement");
                    String data = substring;
                    int lineNumber = Integer.parseInt(data.substring(0, data.indexOf(',')));
                    data = data.substring(data.indexOf(',') + 1);
                    int executionCount = Integer.parseInt(data.substring(0, data.indexOf(',')));
                    data = data.substring(data.indexOf(',') + 1);
                    boolean unexecutedBlock = Integer.parseInt(data) == 1;
                    CoverageFunctionData functionData = currentFile.functionFromLine(lineNumber);
                    functionData.emplaceLine(lineNumber, executionCount, unexecutedBlock);
                }
                break;
                case "branch": {

                }
                break;
                default:
                    break;
            }
        }
    }

    @Override
    public void run() {
        m_data.clearCoverage();
        List<Path> gcdaFiles = new ArrayList<>();
        try {
            Files.find(Paths.get(m_buildDirectory), Integer.MAX_VALUE,
                    ((path, basicFileAttributes) -> basicFileAttributes.isRegularFile() && path.toString().endsWith("gcda")))
                    .forEach(gcdaFiles::add);
        } catch (IOException e) {
            e.printStackTrace();
        }

        for (Path gcda : gcdaFiles) {
            if (!generateGCDA(gcda) || !Paths.get(gcda.toString() + ".gcov").toFile().exists()) {
                continue;
            }

            List<String> lines = new ArrayList<>();
            try (Stream<String> stream = Files.lines(Paths.get(gcda.toString() + ".gcov"))) {
                stream.forEach(lines::add);
            } catch (IOException e) {
                e.printStackTrace();
            }

            if (!Paths.get(gcda.toString() + ".gcov").toFile().delete() || lines.isEmpty()) {
                continue;
            }
            parseGCov(lines);
        }

        for(Path gcda : gcdaFiles) {
            if(!gcda.toFile().delete()) {
                log.warn(gcda.toString() + " could not be deleted");
            }
        }

        if(m_runner != null) {
            ApplicationManager.getApplication().invokeLater(m_runner);
        }
    }
}

File: src/main/kotlin/gcov/data/CoverageThread.kt
package gcov.data

import com.intellij.notification.NotificationType
import com.intellij.notification.Notifications
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.jetbrains.cidr.cpp.toolchains.CPPToolSet
import com.jetbrains.cidr.cpp.toolchains.CPPToolchains
import gcov.notification.GCovNotification
import gcov.state.GCovSettings
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.nio.file.Files
import kotlin.streams.toList

class CoverageThread(private val myProject: Project, private val myBuildDirectory: String,private val toolchain: CPPToolchains.Toolchain, private val myRunner: Runnable?) : Thread() {
    private val myData = CoverageData.getInstance(myProject)
    private val myLinesSaid = HashSet<List<String>>()

    init {
        name = "CoverageThread"
    }

    private fun generateGCDA(gcda: File):Boolean {
        try {
            var lines = mutableListOf<String>()
            val nullFile = if (System.getProperty("os.name").startsWith("Windows")) {
                "NUL"
            } else {
                "/dev/zero"
            }
            val path = GCovSettings.getInstance().getGCovPathForToolchain(toolchain)?.gcovPath
            if(path == null) {
                val notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                        .createNotification("No GCov specified for toolchain $toolchain", NotificationType.ERROR)
                Notifications.Bus.notify(notification, myProject)
                return false
            }
            val builder = if (toolchain.toolSetKind == CPPToolSet.Kind.WSL){
                ProcessBuilder(toolchain.toolSetPath,"run",path, "-i", "-m", "-b", gcda.name.toString())
            }else{
                ProcessBuilder(path, "-i", "-m", "-b", gcda.name.toString())
            }.run {
                directory(gcda.parentFile)
                redirectOutput(File(nullFile))
                redirectErrorStream(false)
            }
            val p = builder.start()
            val reader = BufferedReader(InputStreamReader(p.errorStream))
            do {
                val line = reader.readLine() ?: break
                lines.add(line)
            }while (true)
            val retCode = p.waitFor()
            if (retCode != 0) {
                log.warn("\"gcov\" returned with error code $retCode")
            }
            if(lines.isNotEmpty()) {
                lines = lines.map {
                    val pathConverted = it.replace('\\','/')
                    pathConverted.substring(1 + pathConverted.indexOf(':',pathConverted.indexOf('/')))
                }.toMutableList()
                if(myLinesSaid.add(lines)) {
                    val notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                            .createNotification("gcov returned following warning:\n" + lines.joinToString("\n"),NotificationType.WARNING)
                    Notifications.Bus.notify(notification,myProject)
                }
            }
        } catch (e: IOException) {
            val notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                    .createNotification("\"gcov\" was not found in system path", NotificationType.ERROR)
            Notifications.Bus.notify(notification, myProject)
        } catch (e: InterruptedException) {
            val notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                    .createNotification("Process timed out", NotificationType.ERROR)
            Notifications.Bus.notify(notification, myProject)
        }
        return true
    }

    private fun parseGCov(lines: List<String>) {
        var functionsSet = false
        val version = GCovSettings.getInstance().getGCovPathForToolchain(toolchain)?.version
        if(version == null) {
            val notification = GCovNotification.GROUP_DISPLAY_ID_INFO
                    .createNotification("No GCov specified for toolchain $toolchain", NotificationType.ERROR)
            Notifications.Bus.notify(notification, myProject)
            return
        }
        val wsl = toolchain.wsl
        var currentFile: CoverageFileData? = null
        loop@ for (line in lines) {
            val substring = line.substring(line.indexOf(':') + 1)
            when (line.substring(0, line.indexOf(':'))) {
                "file" -> {
                    currentFile = myData.getCoverageFromPath(wsl?.toLocalPath(null,substring)?.replace('\\','/')
                            ?: substring)
                    if (currentFile == null) {
                        currentFile = if (wsl == null) {
                            CoverageFileData(substring)
                        } else {
                            CoverageFileData(wsl.toLocalPath(null,substring).replace('\\','/'))
                        }
                        myData.setCoverageForPath(currentFile.filePath, currentFile)
                    }
                }
                "function" -> {
                    if (currentFile == null) {
                        log.assertTrue(false, "\"function\" statement found before a file statement")
                        break@loop
                    }
                    val list = substring.split(',')
                    //We are currently skipping execution count when reading
                    functionsSet = false
                    if(version == 8) {
                        val startLine = Integer.parseInt(list[0])
                        val endLine = Integer.parseInt(list[1])
                        val function = list.drop(3).fold("") {lhs,rhs -> if(lhs.isEmpty())rhs else "$lhs,$rhs" }
                        currentFile.emplaceFunction(startLine, endLine, function)
                    } else {
                        val startLine = Integer.parseInt(list[0])
                        val function = list.drop(2).fold("") {lhs,rhs -> if(lhs.isEmpty())rhs else "$lhs,$rhs" }
                        currentFile.emplaceFunction(startLine,-1,function)
                    }
                }
                "lcount" -> {
                    if (currentFile == null) {
                        log.assertTrue(false, "\"lcount\" statement found before a file statement")
                        break@loop
                    }
                    if(!functionsSet) {
                        functionsSet = true
                        if(version < 8) {
                            val list = currentFile.functionData.values.sortedBy { it.startLine }.toList()
                            list.mapIndexed{index,data ->
                                data.endLine = list.getOrNull(index+1)?.startLine?.minus(1) ?: Int.MAX_VALUE
                            }
                        }
                    }
                    val list = substring.split(',')

                    val lineNumber = Integer.parseInt(list[0])
                    val executionCount = Integer.parseInt(list[1])
                    val functionData = currentFile.functionFromLine(lineNumber)
                    functionData.emplaceLine(lineNumber, executionCount)
                }
                "branch" -> {

                }
                "version" -> {

                }
                else -> {
                    log.warn("File is not complete, found symbol " + line.substring(0, line.indexOf(':')))
                    break@loop
                }
            }
        }
    }

    override fun run() {
        synchronized(myData) {
            myData.clearCoverage()
            File(myBuildDirectory).walkTopDown().asSequence().filter {
                it.isFile && when {
                    it.name.endsWith(".gcda") -> true
                    it.name.endsWith(".gcov") -> {
                        it.delete()
                        false
                    }
                    else -> false
                }
            }.toList().forEach{
                if(!generateGCDA(it)) {
                    return@run
                }
            }

            File(myBuildDirectory).walkTopDown().asSequence().filter {
                it.isFile && it.name.endsWith(".gcov")
            }.forEach forEach@{
                val lines = Files.lines(it.toPath()).use { stream ->
                    stream.toList()
                }
                if (!it.delete() || lines.isEmpty()) {
                    return@forEach
                }
                parseGCov(lines)
            }
        }

        if (myRunner != null) {
            ApplicationManager.getApplication().invokeLater(myRunner)
        }
    }

    companion object {

        private val log = Logger.getInstance(CoverageThread::class.java)
    }
}
